packages/drivers/odsp-driver/src/odspDriverUrlResolver2.ts

```diff
@@ -0,0 +1,174 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IRequest } from "@fluidframework/core-interfaces";
+import { IResolvedUrl, IUrlResolver } from "@fluidframework/driver-definitions";
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { getLocatorFromOdspUrl, storeLocatorInOdspUrl, encodeOdspFluidDataStoreLocator } from "./odspFluidFileLink";
+import { resolveDataStore } from "./resolveDataStore";
+import { IOdspResolvedUrl, OdspDocumentInfo, OdspFluidDataStoreLocator } from "./contracts";
+import { createOdspCreateContainerRequest } from "./createOdspCreateContainerRequest";
+import { createOdspUrl } from "./createOdspUrl";
+import { resolveRequest } from "./odspDriverUrlResolver";
+import { getShareLink } from "./graph";
+import { IdentityType, SharingLinkTokenFetcher } from "./tokenFetch";
+
+export class OdspDriverUrlResolver2 implements IUrlResolver {
+    private readonly shareLinkMap = new Map<string, Promise<string>>();
+
+    public constructor(
+        private readonly getSharingLinkToken: SharingLinkTokenFetcher,
+        private readonly identityType: IdentityType = "Enterprise",
+        private readonly logger?: ITelemetryLogger,
+        private readonly appName?: string,
+    ) { }
+
+    public createCreateNewRequest(siteUrl: string, driveId: string, filePath: string, fileName: string) {
+        return createOdspCreateContainerRequest(siteUrl, driveId, filePath, fileName);
+    }
+
+    private getKey(resolvedUrl: IOdspResolvedUrl): string {
+        return `${resolvedUrl.siteUrl},${resolvedUrl.driveId},${resolvedUrl.itemId}`;
+    }
+
+    /**
+     * Takes an already generated data store url (from requestUrl) and appends a path to the
+     * existing data store information
+     * @param requestUrl
+     * @param pathToAppend
+     */
+    public appendDataStorePath(requestUrl: URL, pathToAppend: string): string | undefined {
+        const fluidInfo = getLocatorFromOdspUrl(requestUrl);
+
+        if (!fluidInfo) {
+            return undefined;
+        }
+
+        fluidInfo.dataStorePath = `${fluidInfo.dataStorePath}/${pathToAppend}`;
+        storeLocatorInOdspUrl(requestUrl, fluidInfo);
+
+        return requestUrl.href;
+    }
+
+    /**
+     * Resolves request URL into driver details
+     */
+    public async resolve(request: IRequest): Promise<IOdspResolvedUrl> {
```

Comment:
Will this be able to resolve "old" URLs too? Is there a chance we might receive a request at this level for the older url format? (I know we'll have to make a change to our odspresolver in our code but that should be fine since we have to make a change anyways to switch this from the odspresolver to odspresolver2)

Reply:
1.) Will we still have app url resolver after this change?
2.) I think we will throw if we are not able to resolve the urls that you are talking about.

Reply:
We'll probably get rid of 95% of is since this does the exact same thing and resolves the same URLs and same getabsolute URL right? I suspect there might be logic we find out we need once we start deleting code but hopefully we can get it all consolidated in this layer.

Hmm, I'm just worried about back compat, however if the API layer/behavior for this one vs odspDriverUrlResolver2 are exactly the same... then there should be no worries! :D

Reply:
Yeah, I haven't changed the resolve logic in any way. So its the same which was in in your repo. So should exactly work in the same way. And also we are having the old resolver too in case we find some issues. So this will be non blocking in any way.